### PR TITLE
Allocate nothing

### PR DESCRIPTION
**Pull request by moozzyk**

### CODE REVIEW

File: src/Microsoft.AspNet.SignalR.Core/Infrastructure/MemoryPoolTextWriter.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved. See License.md in the project root for license information.
+
+using System;
+using System.Text;
+using System.IO;
+
+namespace Microsoft.AspNet.SignalR.Infrastructure
+{    public class MemoryPoolTextWriter : TextWriter
```
Comment (davidfowl) [inline src/Microsoft.AspNet.SignalR.Core/Infrastructure/MemoryPoolTextWriter.cs]: Formatting

File: src/Microsoft.AspNet.SignalR.Core/Transports/ForeverTransport.cs [hunk 2]
```diff
@@ -298,16 +309,21 @@ private static Task PerformSend(object state)
...
+                writer.Flush();
+
+                context.Transport.Context.Response.Write(writer.Buffer);
+            }
 
             return TaskAsyncHelper.Empty;
         }
 
-        private class ForeverTransportContext
+        private struct ForeverTransportContext
```
Comment (davidfowl) [inline src/Microsoft.AspNet.SignalR.Core/Transports/ForeverTransport.cs]: Change this back to a class. I'm not sure why I made it a struct, seems like it gets boxed anyways.

File: src/Microsoft.AspNet.SignalR.Core/Transports/ForeverTransport.cs [hunk 3]
```diff
@@ -316,7 +332,7 @@ public ForeverTransportContext(ForeverTransport foreverTransport, object state)
             }
         }
 
-        private class SubscriptionDisposerContext
+        private struct SubscriptionDisposerContext
```
Comment (davidfowl) [inline src/Microsoft.AspNet.SignalR.Core/Transports/ForeverTransport.cs]: Same here

File: src/Microsoft.AspNet.SignalR.Core/Transports/ForeverTransport.cs [hunk 4]
```diff
@@ -333,7 +349,7 @@ public void Set()
             }
         }
 
-        internal class RequestLifetime
+        internal struct RequestLifetime
```
Comment (davidfowl) [inline src/Microsoft.AspNet.SignalR.Core/Transports/ForeverTransport.cs]: Same here

File: src/Microsoft.AspNet.SignalR.Core/Transports/ServerSentEventsTransport.cs [hunk 5]
```diff
@@ -84,19 +91,15 @@ private static Task WriteInit(ServerSentEventsTransport transport)
...
-            transport.OutputWriter.WriteLine();
-            transport.OutputWriter.WriteLine();
-            transport.OutputWriter.Flush();
+            transport.Context.Response.Write(new ArraySegment<byte>(_dataInitialized));
 
             return transport.Context.Response.Flush();
         }
 
-        private class SendContext
+        private struct SendContext
```
Comment (davidfowl) [inline src/Microsoft.AspNet.SignalR.Core/Transports/ServerSentEventsTransport.cs]: Same here

File: src/Microsoft.AspNet.SignalR.Core/Transports/WebSocketTransport.cs [hunk 6]
```diff
@@ -194,10 +197,10 @@ private void OnSocketError(Exception error)
             Trace.TraceError("OnError({0}, {1})", ConnectionId, error);
         }
 
-        private class WebSocketTransportContext
+        private struct WebSocketTransportContext
```
Comment (davidfowl) [inline src/Microsoft.AspNet.SignalR.Core/Transports/WebSocketTransport.cs]: here too

File: tests/Microsoft.AspNet.SignalR.Client.Store.Tests/Microsoft.AspNet.SignalR.Client.Store.Tests.csproj [hunk 7]
```diff
@@ -24,6 +24,7 @@
     <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\..\</SolutionDir>
     <RestorePackages>true</RestorePackages>
     <NuGetPackageImportStamp>b44172aa</NuGetPackageImportStamp>
+    <PackageCertificateThumbprint>38E5E0EDE9521816D1F22973A006919ED1B4A41A</PackageCertificateThumbprint>
```
Comment (davidfowl) [inline tests/Microsoft.AspNet.SignalR.Client.Store.Tests/Microsoft.AspNet.SignalR.Client.Store.Tests.csproj]: ?

Comment (moozzyk) [inline tests/Microsoft.AspNet.SignalR.Client.Store.Tests/Microsoft.AspNet.SignalR.Client.Store.Tests.csproj]: https://github.com/SignalR/SignalR/commit/d41d7526e151c37dc17dba836ac6cfd69fe6ecc7#commitcomment-12890137

**Pull request closed without merging**